blockprotocol/blockprotocol
# [SITE] User API key generation, listing, validation

## What this does
Introduces an `ApiKey` model to generate user API keys.

An API key is in the following format - `b10ck5.[publicId].[privateId]` - where:
- `b10ck5` is a constant prefix 
  - it allow users and repo-scanning tools to identify BP API keys (where it appears as the prefix of a string which otherwise matches the format)
- `publicId` is 16 random bytes encoded in hex to produce 24 characters:
  - it is used to identify API keys when presented by users
  - it is also used to identify API keys when displaying them to users, along with any display name
  - in future it can be used to identify specific API keys users want to revoke, rename, etc
- `privateId` is a uuid (36 characters including dashes)

This format [inspired by Dynatrace's](https://www.dynatrace.com/news/blog/further-enhance-security-by-easily-automating-your-api-token-protection/).

The full API key is only known at the point of creation, with only the `publicId` and `salt` persisted to the database, along with the hash of `[publicId].[privateId].[salt]`.

When a user presents an API key, the key is retrieved by `publicId` and the hashes compared.

For MVP, users can only have a single active key at a time. Generating a new key will revoke any keys associated with the user that have not already been revoked.

We could probably do without a salt for this but I don't think it does any harm, either.

I've added a `searchStub` API endpoint as an example endpoint which requires a valid API key - this will be replaced with the logic in #106 when we pull this together.

## How to test it
Log in to `http://localhost:3000` in the browser, and then either run the following in the console or get your cookie and do it via some other tool.

1. Generate an API key
```JavaScript
fetch("http://localhost:3000/api/me/generateApiKey", { 
    headers: { "content-type": "application/json" }, 
    method: "POST", 
    body: JSON.stringify({ displayName: "test" }) 
 }).then(resp => resp.json()).then(console.log)
```
make a note of the key responded to use in future tests.

2. List keys
```javascript
fetch("http://localhost:3000/api/me/apiKeys").then(resp => resp.json()).then(console.log)
```

3. Try using a key in the dummy endpoint
```javascript
fetch("http://localhost:3000/api/searchStub", { 
  headers: { "x-api-key": "b10ck5.xxxx" } 
}).then(resp => resp.json()).then(console.log)
```
Try invalidly formatted keys, or no key.

4. Generate a new key, then try using the one you generated previously - it should fail.
